### DISCUSSION TITLE
Unifying managed tags and individual tags

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#1846] [Ideas]
Let us discuss the unification of managed tags and ad-hoc tags. This feature has been requested by @euu2021 and supported by @macmarrum, and it is actually possible and there are some open questions.

Currently, we have managed tags and ad-hoc tags. The user can select the tag type by the separator with defaults `::` for managed tags and `--` for categorized ad-hoc tags. The managed tags are saved in nodes without their categories. The classification is managed separately and saved per map. The ad-hoc tags, including their categories, are saved at the node, and there is no management UI for them.

The unification could look as follows:

1. The categories managed in the category manager are applied to both managed tags and ad-hoc tags. Therefore, each time the categories are saved (e.g., "ok" in the category manager is pressed), Freeplane performs a search-and-replace across the whole mind map so that all changes are applied to all tags saved in the nodes, namely to both managed and ad-hoc tags if applicable.
2. Instead of using different separators for managed and ad-hoc tags, they could use the same separator. In this case, we could have a flag defining if tag categories are saved at the map level (as currently with managed tags) or at the node level. The flag could be set in the tag editor for each edited tag separately. However, I am not sure if using the same separator and a checkbox is more convenient than the current solution.
3. Because it would allow the use of the tag category manager for ad-hoc tags, calling them "managed" and "ad-hoc" seems to be no longer adequate. The actual difference is where the categories are saved, at the node level or at a central place only. The best names I found so far could be "managed tags" and "individual tags."

Please let me know what you think so that I could work on the implementation on the coming Saturday.

**Comment by macmarrum** [comment]
Dimitry, thanks for the proposal. It's great to have the topic being discussed further.

I'd like to better understand the benefits (or at least differences) between "managed tags" and "individual tags". What functionality (filtering, searching – via UI, via the scripting API) is supported for each of the models?

| |`Edit->Find->Find…`|`Filter->Compose filter`| API, e.g. `c.find { it.tags.tags.any { it.startsWith('food::fruit::') } }`|
|-|-|-|-|
|managed - category::tag | y/n? | y/n? | y/n? |
|managed - tag only | y/n? | y/n? | y/n? |
|individual - category::tag | y/n? | y/n? | y/n? |
|individual - tag only | y/n? | y/n? | y/n? |

Are there any other benefits/differences that you can think of?

**Reply by dpolivaev** [reply]
You can assume "y" everywhere, it can be implemented for both kinds of tags.

There are two differences: 

1. Individual tags allow to differentiate between equally named tags belonging to different categories, and managed tags are automatically appearing in all categories if they occur in the category tree multiple times.
2. Individual tags are much harder to implement because each modification of the category tree e.g. drag and drop or category tag rename should be mapped to the changes in the nodes.

**Reply by macmarrum** [reply]
I see. Thanks, Dimitry.

Once "individual tag" categories can be reordered by drag-and-drop and renamed, will anyone want to use the Schrödinger's-cat-style "managed tags" model?

In other words, I am struggling to come up with viable use cases for the "managed tags" model as compared to the "individual tags".

Let's take the example of `tomato` which can be in both `food::vegetable::tomato` and `food::fruit::tomato`.
Using "managed tags" I add a single tag `tomato` with the meaning of both `food::vegetable::tomato` and `food::fruit::tomato`.
Using "individual tags" I need to add both tags `food::vegetable::tomato` and `food::fruit::tomato` to achieve the same. But when I want to find the `fruit` variant of `tomato`, only "individual tags" support that.

So it seems to me that the benefit of "managed tags" is to have `tomato` displayed only ones in the node, whereas with "individual tags" there are two `tomato` tags displayed. And I'm wondering if this warrants keeping the "managed tags" model.

**Reply by dpolivaev** [reply]
Managed tags represent "one tag can belong to multiple categories" concept.
For example, `Benjamin Franklin` - he is known both as a foundational political figure in American history and as a scientist. He was a philosopher, political theorist, and inventor alongside his scientific pursuits which included studies in electricity, meteorology, and oceanography. Franklin's role in the development of the American system of government is as celebrated as his scientific experiments and inventions, like the lightning rod.
Individual tags mean that tags associated with different categories should not be mixed.

**Reply by dpolivaev** [reply]
I think that managed tags are closer to the book index where same term appears at different pages or in different chapters.

**Reply by dpolivaev** [reply]
The tags are displayed only one at the node. However when you open the tag editor that shows the tags qualified with their categories, managed tags are be shown multiple times if they belong to multiple categories.

**Reply by dpolivaev** [reply]
My personal preference to managed tags comes not from the use case but from "don't repeat yourself" principle because in my eyes duplication of categories at the nodes is data duplication and as a software developer I want to eliminate duplication whenever possible.

**Reply by macmarrum** [reply]
Given both category models are kept, I like the idea of using the same separator and a flag to distinguish the models.

The way I understand the proposed solution is that "managed tags" will be saved in 2 places:
1. in each node which is tagged
2. in the mind map, as an index of categorized tags

whereas "individual tags" will be saved only in each node that is tagged.

E.g. when I add a new "managed tag" `food::vegetable::tomato` to node A, and `food::fruit::tomato` to node B, then in both nodes A and B only `tomato` is stored, but in the index, both `food::vegetable::tomato`, and `food::fruit::tomato` is added and stored.

This is different from "individual tags". In this model `food::vegetable::tomato` is stored in node A, and `food::fruit::tomato` in node B.

I think "managed tags" could be called "**indexed tags**", and "individual tags" – "**stand-alone tags**".

The flag could be called "index" or "is indexed".

**Reply by dpolivaev** [reply]
If you add a new line `food::vegetable::tomato` to the tag editor, how should it decide if the new tag is managed or individual?

**Reply by dpolivaev** [reply]
All tag categories are indexed, therefore "indexed" is not a good fit

**Reply by macmarrum** [reply]
> All tag categories are indexed, therefore "indexed" is not a good fit

Maybe "partitioned" would be a better description for "managed tags". And "individual tags" could be called "full-length".

> If you add a new line food::vegetable::tomato to the tag editor, how should it decide if the new tag is managed or individual?

A good point. And also when choosing from a drop-down list of existing tags.
I can see a flag bring in additional challenges. So maybe let's stay with two separators.

I'd only propose to change the default for "full-length" aka "individual tags" from `--` to `..`.
I like better the look of `food..vegetable..tomato` than `food--vegetable--tomato`. Also `..` is easier to type for me than `--`.

**Reply by dpolivaev** [reply]
> And also when choosing from a drop-down list of existing tags.

Do you mean that all tags should appear in the drop-down list twice? Probably it would also be not the best solution.

Actually, if we assume that people have a general preference to always use one or another approach and do not mix them, a single checkbox could help to resolve it.

**Reply by macmarrum** [reply]
In case of a separator-based segregation between **partitioned** and **full-length** tags, this is how I imagine it:

Given two "full-length" tags: `name..Wednesday` and `workday..Wednesday`, and a "partitioned" tag `Wednesday` belonging to the categories `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I open a drop-down in Tags Editor
Then I can see the following entries
- `name..Wednesday`
- `workday..Wednesday`
- `letter::w::Wednesday`
- `mythology::Odin::Wednesday`

In case of a flag-base segregation, I think that a checkbox in Tag Editor will also be fine with me.

Given two "full-length" tags: `name::Wednesday` and `workday::Wednesday`, and a "partitioned" tag `Wednesday` belonging to the categories `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I open a drop-down in Tags Editor and have the checkbox `[x] Partitioned` enabled
Then I can see the following entries
- `letter::w::Wednesday`
- `mythology::Odin::Wednesday`

Given two "full-length" tags: `name::Wednesday` and `workday::Wednesday`, and a "partitioned" tag `Wednesday` belonging to the categories `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I open a drop-down in Tags Editor and have the checkbox `[ ] Partitioned` disabled
Then I can see the following entries
- `name::Wednesday`
- `workday::Wednesday`

**Reply by dpolivaev** [reply]
> Given two "full-length" tags: name..Wednesday and workday..Wednesday, and a "partitioned" tag Wednesday belonging to the categories letter::w::Wednesday, mythology::Odin::Wednesday

Do you think that we need separate category trees for the full-length tags and for the partitioned tags? I think it is too complicated for the user. And it means that we need the checkboxes.

**Reply by macmarrum** [reply]
> Do you think that we need separate category trees for the full-length tags and for the partitioned tags?

Yes. In Tag Category Manager I'd like to manage both partitioned and full-length tags, and to be able to tell them apart. It's fine with me to have two trees, one under another.

> I think it is too complicated for the user. And it means that we need the checkboxes.

It's fine if only one tree (either partitioned or full-length) is displayed at a time, depending on a checkbox.

**Reply by dpolivaev** [reply]
Can you describe a use case that requires two different category trees in the single mind map? In my eyes it is YAGNI. Nobody can ever need it if I am right.

**Reply by macmarrum** [reply]
Merging information from two mind maps, each using a different tag model.

Adam uses full-length tags in his mind maps. He creates a mind map and shares it with his team.
Eve prefers partitioned tags in her mind maps. She needs to incorporate some of the information form Adam's mind map, so she does a copy-paste into her mind map. The original nodes are full-length-tagged. When she opens Tag Category Manager, she needs to see both tag-category models.

**Reply by dpolivaev** [reply]
I am sorry, I can not imagine it. But thank you for the discussion.

I am going to keep only one separator and have only one category tree per map.
I am going to have a checkbox "save categories with tags" that influences if the categories are saved at the nodes.
I am going to save the checkbox value (true or false) with the mind map when the tags are saved.

When nodes containing tags with categories are copied, I am going to add all categories to the map category tree.

**Reply by dpolivaev** [reply]
I have considered the following alternatives:

- Having this checkbox for each tag individually
- Having different separators

I think they are not as easy to understand and use as the option I chose.

**Reply by macmarrum** [reply]
That's fine with me.

> I am going to keep only one separator and have only one category tree per map.
I am going to have a checkbox "save categories with tags" that influences if the categories are saved at the nodes.
I am going to save the checkbox value (true or false) with the mind map when the tags are saved.

I'm interested to learn how the proposed solution will handle a change of "save categories with tags" during the lifetime of a mind map:

Given nodes tagged with "full-length" tags: node A with `name::Wednesday`, node B with `workday::Wednesday`
When I disable "save categories with tags"
Then ...?

Given node A and B tagged with "partitioned" tags `Wednesday`, and the index containing `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I enable "save categories with tags"
Then ...?

> When nodes containing tags with categories are copied, I am going to add all categories to the map category tree.

I'm also interested about how the following will be handled as far as tags/categories are concerned:

Given node A and B tagged with a "partitioned" tag `Wednesday`, and the index containing `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I copy node A and B and paste it to my other mind map with "save categories with tags" enabled
Then ...?

**Reply by dpolivaev** [reply]
When you disable "save categories with tags", the categories are not saved, but they are always added to the category tree. Therefore the saved tags can be found under the both categories, and the both categories should be displayed in the tag editor when you open it again.

If categories are not saved with the tags, they are not copied / pasted with the nodes and with the tags. In this case the category tag of the target map are not modified.

**Reply by macmarrum** [reply]
> When you disable "save categories with tags", the categories are not saved, but they are always added to the category tree. Therefore the saved tags can be found under the both categories, and the both categories should be displayed in the tag editor when you open it again.

From your description I understand the following

Given nodes tagged with "full-length" tags: node A with `name::Wednesday`, node B with `workday::Wednesday`
When I disable "save categories with tags"
Then **in node A `name::Wednesday` is replaced with `Wednesday`, in node B `workday::Wednesday` is replaced with `Wednesday`, and both `name::Wednesday` and `workday::Wednesday` are added to the category tree**

Please let me know if it isn't what you had in mind.

And what about the other way round?

Given node A and B tagged with "partitioned" tags `Wednesday`, and the category tree contains `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I enable "save categories with tags"
Then ...?

> If categories are not saved with the tags, they are not copied / pasted with the nodes and with the tags. In this case the category tag of the target map are not modified.

This is how I translate the description to the **Then** section:

Given node A and B tagged with "partitioned" tags `Wednesday`, and the index containing `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I copy node A and B and paste it to my other mind map with "save categories with tags" enabled
Then **both nodes A and B have a (category-less) full-length tag `Wednesday`**

I'd propose to always copy categorized tags and treat them on paste according to the target tag-category model. This would change the **Then** section to: 

**both nodes A and B have two full-length tags: `letter::w::Wednesday`, `mythology::Odin::Wednesday`**

What do you think about it, Dimitry?

**Reply by dpolivaev** [reply]
> both nodes A and B have two full-length tags: letter::w::Wednesday, mythology::Odin::Wednesday

yes, it is the idea

**Reply by macmarrum** [reply]
OK, great.

Given-when-then `#2` is similar in nature to `#3`, so I think `#2` will have the same effect (Then)

Given node A and B tagged with "partitioned" tags `Wednesday`, and the category tree contains `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I enable "save categories with tags"
Then **both nodes A and B have two full-length tags: `letter::w::Wednesday`, `mythology::Odin::Wednesday`**

Agreed?

**Reply by macmarrum** [reply]
EDIT: so I think `#2` will have the same effect (Then)

**Reply by dpolivaev** [reply]
For me the index means the same as the category tree (at the time point when the editor was opened, I am not going to update editors in case of any changes in between)

**Reply by macmarrum** [reply]
> For me the index means the same as the category tree

That's also how I understand it.

> I am not going to update editors in case of any changes in between

Understood. It makes sense to me.

The above things aside, is given-when-then `#2` fine with you?

Given node A and B are tagged with a "partitioned" tag `Wednesday`
 and the category tree (aka index) contains `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I enable "save categories with tags"
Then **both nodes A and B have two full-length tags: `letter::w::Wednesday`, `mythology::Odin::Wednesday`**

**Reply by dpolivaev** [reply]
Seeing a lot of similar looking texts I do not see the difference. Wait until I implement it.

**Reply by macmarrum** [reply]
> Seeing a lot of similar looking texts I do not see the difference.

Difference in what? Sorry, you lost me there.

> Wait until I implement it.

As you wish.

Please keep in mind that you had asked for more precise description of our needs and use cases so that, as I understand, you could avoid misunderstanding and doing "unneeded" work → "_I’ve spent a significant part of my vacation working on the tag category system, only to learn that you find little use for it. I’m sorry to hear that. It appears I developed the system based on assumptions that seemed logical to me, without sufficient user use cases and examples._"
So if we stop the discussion before having confidence that both sides understand each other, then we run the risk of repeating the earlier situation.

**Reply by dpolivaev** [reply]
I just do not understand what you mean. You lost me too in your questions.

**Reply by macmarrum** [reply]
OK. Let me put it as blunt questions.

This is my use case. 

`#2`
Given node A and B are tagged with a "partitioned" tag `Wednesday`
 and the category tree (aka index) contains `letter::w::Wednesday`, `mythology::Odin::Wednesday`
When I enable "save categories with tags"
Then **both nodes A and B have two full-length tags: `letter::w::Wednesday`, `mythology::Odin::Wednesday`**

1. Do you understand it?
2. Will you attempt to implement it?

**Reply by dpolivaev** [reply]
1: I think, I do. 2: yes.

**Comment by dpolivaev** [comment]
How should Freeplane modify qualified tags is their categories or their parent categories are deleted in the category manager?

- Should such tags be removed?
- Or should only their categories be removed?

**Reply by dpolivaev** [reply]
If the categories are renamed, dragged and dropped oder cut and pasted, I am going to rename categories saved in the nodes and keep the tags. Actually, I think it is the most difficult to implement logic. I am writing it here to make sure we are aligned. If not, please comment.

**Reply by macmarrum** [reply]
> How should Freeplane modify qualified tags is their categories or their parent categories are deleted in the category manager?
>
>Should such tags be removed?
Or should only their categories be removed?

I would use a directory/file analogy. If I remove a directory, all files in that directory are removed. But before removing a directory, a file manager usually asks for confirmation "Really remove the directory and all the files it contains?" or similar.

**Reply by dpolivaev** [reply]
Currently when you edit categories (or tags), the action is applied only when you click on OK, but you always can click on CANCEL too. Therefore I am preparing the list of replacements / deletions to be executed on OK. I do not understand when the question can be asked and how to react if you say "no".

**Reply by dpolivaev** [reply]
We should also always consider cut and paste. If you cut, nobody knows if you paste it back until you finish the editing.

**Reply by macmarrum** [reply]
> If the categories are renamed, dragged and dropped oder cut and pasted, I am going to rename categories saved in the nodes and keep the tags. Actually, I think it is the most difficult to implement logic. I am writing it here to make sure we are aligned. If not, please comment.

Yes. I think it covers my use case:

Given 4 nodes, each tagged with 1 fully-quallified tag: `food::vegetable::tomato`, `food::vegetable`, `food::vegetableX::tomato`, `fridge::food::vegetable`
When User renames `vegetable` in `food::vegetable`
Then `food::vegetable::tomato` is replaced with `food::fruit::tomato` and `food::vegetable` becomes `food::fruit`, and the other 2 tags remain unchanged

**Reply by macmarrum** [reply]
> Currently when you edit categories (or tags), the action is applied only when you click on OK, but you always can click on CANCEL too. Therefore I am preparing the list of replacements / deletions to be executed on OK. I do not understand when the question can be asked and how to react if you say "no".
> We should also always consider cut and paste. If you cut, nobody knows if you paste it back until you finish the editing.

Given the category tree
* food
  * vegetable
    * tomato
    * cucumber
* fridge
  * fruit
    * strawberries

When I remove or cut `vegetable`
Then I get a confirmation dialog "The category is not empty. If you remove it, all its children will be removed too"

**Comment by Illegit** [comment]
I add a topic to this thread: please remember to create a command to delete all of the tags of the node: like the remove command for notes, details, and so on. Thanks.

**Comment by euu2021** [comment]
The Benjamin Franklin example made me notice the power of managed tags. For example, in this situation:

The tag Benjamin Franklin can be categorized like this:
```
political figures::Benjamin Franklin
scientists::Benjamin Franklin
philosophers::Benjamin Franklin
```

Let's say I have a node representing a certain scientific concept, for example "electricity", and I want to tag it with the people who helped develop this idea. So, I tag it with `Benjamin Franklin`.

Later, I want to find all the scientific concepts that were influenced by philosophers. Using managed tags, it would be easy, because I can simply use, as the search criteria, the category `philosophers`. On the other hand, with individual tags, it would be a problem, because, when tagging the node, I would need to have included all the 3 tags, which is not practical.

So, it seems that the power of the managed tags is at the moment of retrieval during a search. If that's true, then I think that there is a way to "unify" the two kinds of tags. In other words, bring the power of managed tags into individual tags. The solution would be to make the search more powerful, so the individual tags also have the powerful retrieval. I explain:

In the Benjamin Franklin example above, I would tag the node "electricity" with any of the tags, for example `scientists::Benjamin Franklin` (or even just use the tag `Benjamin Franklin` without any category). Then, when I want to find the scientific concepts influenced by philosophers, I would need to have an option to use an "expanded search criteria", that would work like this:

- the user chose the category `philosophers`;
- so, to create the "expanded search criteria", FP identifies every tag inside the category `philosophers`. In the example, it's only the tag Benjamin Franklin;
- then, FP searches for every other category that Benjamin Franklin belongs to, and include those other tags in the search criteria, creating the "expanded search criteria". In the example, it would include `political figures::Benjamin Franklin` and `scientists::Benjamin Franklin`;
- then it searches for the nodes that match the "expanded search criteria". In the example, it would successfully find the "electricity" node, because it has the tag `scientists::Benjamin Franklin`, which is included in the "expanded search criteria".

So, with this "expanded search criteria" option, it would be possible to achieve, with individual tags, the same result that is achieved with managed tags, what would make it possible to remove the managed tags feature.

Does that make sense?

**Reply by dpolivaev** [reply]
I love it.

However, what should Freeplane do if someone just enters "Benjamin Franklin" as a tag value?

**Reply by euu2021** [reply]
> However, what should Freeplane do if someone just enters "Benjamin Franklin" as a tag value?

Then, the tag tree, at the moment of the search, would be like this:

```
political figures::Benjamin Franklin
scientists::Benjamin Franklin
philosophers::Benjamin Franklin
Benjamin Franklin
```

So, the standalone `Benjamin Franklin` would also be included in the "expanded search criteria". In other words, every entry of the tag tree that includes `Benjamin Franklin` would be included in the "expanded search criteria".

**Reply by dpolivaev** [reply]
Makes sense.

**Comment by Illegit** [comment]
And even consider to include in that search the eventual category tags as:

> Inventors::Benjamin Frankilin::Flashes
> Politcs::Political figures::Benjamin Frankilin::US Constitution

**Comment by dpolivaev** [comment]
Preview 1.12.2-01 alpha
Now tag category editor supports also qualified tags in the nodes, and category search considers all categories for each tag.
There is no warning for deletion / replacements of tags because I do not know where, when  and how to display it. I have also not changed wording used in the UI. Your ideas and constructive feedback  are appreciated.

All in all it was quite a difficult thing.

**Reply by euu2021** [reply]
Now, the logic is ideal, for me. Thanks!

**Reply by macmarrum** [reply]
I like that there is just one type of tags (one tag separator `::`), that all tags are actually fully-qualified and that they can be managed in **Tag category manager**.
I think it's great that the functionality of what used to be "partitioned" tags is now provided as "extended search" :heart:

**Comment by Illegit** [comment]
+1. Compliments for last tags solution.
